crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1072,6 +1079,33 @@ impl AuthorityPerEpochStore {
         Ok(())
     }
 
+    pub fn are_consensus_messages_processed<'a>(
+        &self,
+        keys: impl Iterator<Item = &'a SequencedConsensusTransactionKey>,
+    ) -> SuiResult<Vec<bool>> {
+        Ok(self
+            .tables
+            .consensus_message_processed
+            .multi_contains_keys(keys)?)
+    }
+
+    pub async fn consensus_messages_processed_notify(
+        &self,
+        keys: Vec<SequencedConsensusTransactionKey>,
+    ) -> Result<(), SuiError> {
+        let unprocessed_keys = keys
+            .iter()
+            .zip(
+                self.are_consensus_messages_processed(keys.iter())?
+                    .into_iter(),
+            )
+            .filter_map(|(key, processed)| if !processed { Some(*key) } else { None })
+            .collect::<Vec<SequencedConsensusTransactionKey>>();
+        let registrations = self.consensus_notify_read.register_all(unprocessed_keys);
```

Comment:
I think you need to register waiters before reading the keys, or there will be a race: a key can be written between `are_consensus_messages_processed` and the time when you register the waiters.

Reply:
Thanks for catching that Mark!